### PR TITLE
extend set_cipher_list() compatibility layer API

### PR DESCRIPTION
**Pull request by miyazakh**
Changes to extend set_cipher_list() compatibility layer API to have set_ciphersuites compatibility layer API capability.
 
 1. the same behavior when list has mixed, pre-TLS13 suites and TLS13 suites
 2. only update pre-TLS13 suites when list has only pre-TLS13 suites
 3. only update TLS13 suites when list has only TLS13 suites

Currently, both set_cipher_list() and set_ciphersuites() are mapped to wolfSSL_set_cipher_list. Therefore, suites are overwritten even if cipher list has only pre-TLSv13 or TLS13.

### CODE REVIEW

File: src/ssl.c [hunk 1]
```diff
@@ -12005,6 +12005,166 @@ static int wolfSSL_remove_ciphers(char* list, int sz, const char* toRemove)
...
+        int    i;
+        word32 length;
+        
+        next   = XSTRSTR(next, ":");
+        length = min(sizeof(name), !next ? (word32)XSTRLEN(current) /* last */
+                                         : (word32)(next - current));
+        XSTRNCPY(name, current, length);
+        name[(length == sizeof(name)) ? length - 1 : length] = 0;
+        
+        for (i = 0; i < suiteSz; i++) {
```
Comment (SparkiDev) [inline src/ssl.c:12142]: Use wolfSSL_get_cipher_suite_from_name().

File: src/ssl.c [hunk 2]
```diff
@@ -12005,6 +12005,166 @@ static int wolfSSL_remove_ciphers(char* list, int sz, const char* toRemove)
...
+        
+        if (XSTRNCMP(enabledcs, "None", XSTRLEN(enabledcs)) != 0) {
+            len += (word32)XSTRLEN(enabledcs) + 2;
+        }
+    }
+    
+    len += listsz + 2;
+    
+    /* build string */
+    if (len > 0) {
```
Comment (SparkiDev) [inline src/ssl.c:12059]: Can this fail given statement above?
Change to compare with 2 I assume.

Also return NULL when nothing to return and then no need to put the construction code in a block.

File: src/ssl.c [hunk 3]
```diff
@@ -12005,6 +12005,166 @@ static int wolfSSL_remove_ciphers(char* list, int sz, const char* toRemove)
...
+            ianasz = (int)XSTRLEN(enabledcs);
+            if (ianasz + 1 < len) {
+                XSTRNCPY(locallist, enabledcs, len);
+                locallist += ianasz;
+                
+                *locallist++ = ':';
+                *locallist = 0;
+                len -= ianasz + 1;
+            }
+            else{
```
Comment (SparkiDev) [inline src/ssl.c:12100]: Cipher suite should have been validated above when calculating length.

**Review by SparkiDev — CHANGES_REQUESTED** [review]

File: src/ssl.c [hunk 4]
```diff
@@ -12005,6 +12005,166 @@ static int wolfSSL_remove_ciphers(char* list, int sz, const char* toRemove)
 
     return totalSz;
 }
+/*                                                                          */
```
[the remark this answers is not in the document]
Comment (miyazakh) [reply src/ssl.c:12008]: Changed

File: src/ssl.c [hunk 5]
```diff
@@ -12005,6 +12005,166 @@ static int wolfSSL_remove_ciphers(char* list, int sz, const char* toRemove)
...
+/* build enabled cipher list w/ TLS13 or w/o TLS13 suites                   */
+/* @param ctx    a pointer to WOLFSSL_CTX structure                         */
+/* @param suites currently enabled suites                                   */
+/* @param onlytlsv13suites flag whether correcting w/ TLS13 suites          */
```
[the remark this answers is not in the document]
Comment (miyazakh) [reply src/ssl.c:12012]: Changed

File: src/ssl.c [hunk 6]
```diff
@@ -12005,6 +12005,166 @@ static int wolfSSL_remove_ciphers(char* list, int sz, const char* toRemove)
...
+    
+    /* calculate necessary buffer length */
+    for(idx = 0; idx < suites->suiteSz; idx++) {
+        
+        cipherSuite0 = suites->suites[idx];
+        cipherSuite  = suites->suites[++idx];
+        
+        if (onlytlsv13suites && cipherSuite0 == TLS13_BYTE) {
+            enabledcs = GetCipherNameInternal(cipherSuite0, cipherSuite);
+        } else if (!onlytlsv13suites && cipherSuite0 != TLS13_BYTE) {
```
[the remark this answers is not in the document]
Comment (miyazakh) [reply src/ssl.c:12046]: Changed

File: src/ssl.c [hunk 7]
```diff
@@ -12068,8 +12231,35 @@ static int wolfSSL_parse_cipher_list(WOLFSSL_CTX* ctx, Suites* suites,
...
+        
+        listattribute = CheckcipherList(list);
+        
+        if (listattribute == 0) {
+            /* list has mixed(pre-TLSv13 and TLSv13) suites    */
+            /* update cipher suites the same as before         */
+            return (SetCipherList(ctx, suites, list)) ? WOLFSSL_SUCCESS :
             WOLFSSL_FAILURE;
+        } else if (listattribute == 1) {
+        /* list has only pre-TLSv13 suites. Only update before TLSv13 suites.*/
```
[the remark this answers is not in the document]
Comment (miyazakh) [reply src/ssl.c:12243]: Moved in by 4 spaces

File: src/ssl.c (hunk 1 above)
Comment (miyazakh) [reply src/ssl.c:12142]: Changed to use wolfSSL_get_cipher_suite_from_name()

File: src/ssl.c (hunk 2 above)
Comment (miyazakh) [reply src/ssl.c:12059]: Thanks. Changed to compare with 2.

File: src/ssl.c (hunk 3 above)
Comment (miyazakh) [reply src/ssl.c:12100]: Hello @SparkiDev 
This means that cipher suites in the "list" that user sets should be validated? I think that it is done in SetCipherList(). Therefore, it just passes the list as is. I should miss something. Please let me know more detail.

File: src/ssl.c [hunk 8]
```diff
@@ -12036,6 +12036,172 @@ static int wolfSSL_remove_ciphers(char* list, int sz, const char* toRemove)
...
+    int flags;
+    char* next = (char*)list;
+    
+    do {
+        char*  current = next;
+        char   name[MAX_SUITE_NAME + 1];
+        word32 length;
+        
+        next   = XSTRSTR(next, ":");
+        length = min(sizeof(name), !next ? (word32)XSTRLEN(current) /* last */
```
Comment (SparkiDev) [inline src/ssl.c:12175]: Make it the min(sizeof(name) - 1, ...) and there will always be space for the '\0'.

File: src/ssl.c [hunk 9]
```diff
@@ -12036,6 +12036,172 @@ static int wolfSSL_remove_ciphers(char* list, int sz, const char* toRemove)
...
+                /* TLSv13 suite */
+                findTLSv13Suites = 1;
+                break;
+            } 
+            else {
+                findbeforeSuites = 1;
+                break;
+            }
+        } 
+        if (findTLSv13Suites == 1 && findbeforeSuites == 1)
```
Comment (SparkiDev) [inline src/ssl.c:12193]: Put a braces around lines in if.
(Less of a maintenance issue.)

File: src/ssl.c [hunk 10]
```diff
@@ -12099,8 +12268,41 @@ static int wolfSSL_parse_cipher_list(WOLFSSL_CTX* ctx, Suites* suites,
...
             WOLFSSL_FAILURE;
+        } 
+        else if (listattribute == 1) {
+            /* list has only pre-TLSv13 suites. 
+             * Only update before TLSv13 suites.
+             */
+            tls13Only = 1;
+        } 
+        else if (listattribute == 2) {
+        /* list has only TLSv13 suites. Only update TLv13 suites
```
Comment (SparkiDev) [inline src/ssl.c:12288]: Move comment in 4 spaces.

**Review by SparkiDev — CHANGES_REQUESTED** [review]

File: src/ssl.c (hunk 10 above)
Comment (miyazakh) [reply src/ssl.c:12288]: Moved comment in 4 spaces

File: src/ssl.c (hunk 9 above)
Comment (miyazakh) [reply src/ssl.c:12193]: Put braces

File: src/ssl.c (hunk 8 above)
Comment (miyazakh) [reply src/ssl.c:12175]: Oh, Thanks.

Comment (miyazakh) [reply src/ssl.c:12175]: Change code to the following causes error "‘__builtin___strncpy_chk’ specified bound depends on the length of the source argument" on gcc-9.3 that I am using as default gcc compiler,

        length = min(sizeof(name) - 1, !next ? (word32)XSTRLEN(current) /* last */
                                         : (word32)(next - current));
        XSTRNCPY(name, current, length);
        name[length] = 0;

 gcc-7 and gcc-8 are ok. Any idea?

Comment (SparkiDev) [reply src/ssl.c:12175]: Not using the ternary operator appears to fix it.
1. Calculate the current length
2. assign sizeof(name) - 1 to length and
3. assign current length to length when current length is smaller

Comment (miyazakh) [reply src/ssl.c:12175]: Thanks!

**Review by SparkiDev — APPROVED** [review]

**Pull request merged**